### ISSUE TITLE
Rule proposal: Do not use this before defining all members

### ISSUE THREAD

**Issue by iliubinskii** [typescript-eslint/typescript-eslint#6206]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Do not use this before defining all members

### Fail Cases

```typescript
interface I {
  x: number;
  y: number;
}

function f(value: I): number {
  return value.x * value.y;
}

class A implements I {
  x: number;

  y: number;

  constructor() {
    this.x = 1;
    f(this); // Error - "this" is passed to a function as a whole before it is fully defined
    this.y = 2;
  }
}
```

### Pass Cases

```typescript
interface I {
  x: number;
  y: number;
}

function f(value: I): number {
  return value.x * value.y;
}

class A implements I {
  x: number;

  y: number;

  constructor() {
    this.x = 1;
    this.y = 2;
    f(this); // Ok - "this" is fully defined
  }
}
```

**Comment by bradzacher** [comment]
A lot of the dumber cases are covered by `strictPropertyInitialization` - for example with the compiler option turned on you can't do `this.y = this.y + 1` or pass `f(this.y)` because TS will error saying "`y` is used before it's defined"

Whilst I do see value in this rule - I do wonder about its broader applicability to the TS community as a whole. Not everyone uses classes to begin with, and fewer still use classes and call functions in a constructor that take the entire class instance.
It's a very, very niche error.

I think that we can leave this open for a little bit to gauge community engagement, but right without engagement it's probably not going to be something we accept here.

**Comment by bradzacher** [comment]
After almost 6 months there has been zero community engagement.
As such we'll take this as signal close this proposal as not broadly applicable.
